<a href="https://colab.research.google.com/github/angelajmzhou/ecs-171/blob/main/Lab1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Lab1: Complete the TODO parts in the following code.
- Using California Housing Dataset from sklearn, select input attributes 1,3,4  as the input features.
- Using K-fold cross validation technique (https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html), complete the implementation to train a regression model and report performance merics when asked in the following code.
- For multiple degrees of model complexity (i.e., degree of polynomial in this exercise) in a for-loop, obtain the model with the minimum reducible_error, polynomial degree, and run the obtained model on the test data. For this part,you should use the split the data into train and test by [75:25] rate and report mse of the final model on test data.
- Analyse the results of model performance according to different degrees of polynomial and the number of folds used. You can manipulate the code and share your analysis in terms of the performance of the model (mse and total error), such as for instnace which degree of the model complexity (in relation to the polynomial order) would give a better model? Feel free to include other analysis about the generated models in relation to their performance results. You can event plot the results to support your analysis.

In [8]:
###### import numpy as np
from sklearn.model_selection import KFold
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import PolynomialFeatures
from sklearn.datasets import fetch_california_housing
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

def polynomial_regression(degree, X, y, folds, test_size=0.25, random_state=None):
    # Define number of folds for cross-validation
    kf = KFold(folds)

    # Initialize lists to store results for variance, bias2s, total_error, and models
    # TODO
    variance_list = []
    bias2s_list = []
    total_error = []
    models = []

    # Set the polynomial degree of the model
    poly_features = PolynomialFeatures(degree)
    X_poly = poly_features.fit_transform(X)

    # Perform cross-validation
    for train_index, test_index in kf.split(X_poly):
        # Split data into training and testing sets for this fold
        X_train, X_test = X_poly[train_index], X_poly[test_index]
        y_train, y_test = y[train_index], y[test_index]

        # Fit polynomial regression model
        model = LinearRegression()
        model.fit(X_train, y_train)

        # Make predictions on the test set
        y_pred = model.predict(X_test)

        # Calculate variance and Bias^2 for this fold
        variance = np.mean((np.mean(y_pred)-(y_pred))**2)
        bias2 = np.mean(((np.mean(y_pred)- y_test)) ** 2)

        # Append results to lists
        variance_list.append(variance)
        bias2s_list.append(bias2)
        total_error.append((bias2+variance))
        models.append(model)

        # Print results  including variance, Bias^2 and the total_error for this fold.
        print("Variance: {:.4f}, Bias2: {:.4f}, Total error: {:.4f}".format(variance, bias2, bias2 + variance))

    # print the total_error of the best model
    min_error_index = np.argmin(total_error)
    best_model = models[min_error_index]
    print("model number: {:1f} total error :{:4f}".format(min_error_index, total_error[min_error_index]))

    # Testing the final model on the test data
    X_train, X_test, y_train, y_test = train_test_split(X_poly, y, test_size=test_size, random_state=42)
    # Obtain the predictions on the test data
    y_pred_test = best_model.predict(X_test)
    test_variance = np.mean(((y_pred)-np.mean(y_pred))**2)
    test_bias2 = np.mean(((np.mean(y_pred)- y_pred_test)) ** 2)
    print("Best Model Values: Variance: {:.4f}, Bias2: {:.4f}".format(test_variance, test_bias2,))
    # Calculate the mean squared error (MSE) on the test data
    mse = mean_squared_error(y_test, y_pred_test)
    #also find error
    return mse , best_model, total_error[min_error_index]

# Example usage: load California Housing Dataset and select the first, third, and forth attributes as input features in X
# TODO
housing_data = fetch_california_housing()
X = housing_data.data[:, [0, 2, 3]] #: means all rows, [0,2,3] means these specific columns
y = housing_data.target

degrees = range(1, 6)  # Try polynomial degrees from 1 to 5
folds_range = range(5, 21, 5) #try fold #'s 5-20
folds_mse_list = []
degree_error_list = []
# Loop over different folds and degrees
for fold in folds_range:
    fold_mse_list = []
    for degree in degrees:
        mse, best_model, total_error = polynomial_regression(degree, X, y, folds=fold, random_state=42)
        degree_error_list.append({"Fold": fold, "Degree": degree, "Total Error": total_error, "MSE": mse})


Variance: 0.6159, Bias2: 1.1592, Total error: 1.7751
Variance: 0.6857, Bias2: 1.2147, Total error: 1.9004
Variance: 0.8001, Bias2: 1.4449, Total error: 2.2450
Variance: 0.4341, Bias2: 1.1848, Total error: 1.6190
Variance: 0.7587, Bias2: 1.4722, Total error: 2.2309
model number: 3.000000 total error :1.618951
Best Model Values: Variance: 0.7587, Bias2: 0.6617
Variance: 1.4212, Bias2: 1.1390, Total error: 2.5602
Variance: 0.6013, Bias2: 1.1999, Total error: 1.8012
Variance: 0.8027, Bias2: 1.4435, Total error: 2.2462
Variance: 0.5257, Bias2: 1.1828, Total error: 1.7085
Variance: 0.8059, Bias2: 1.4748, Total error: 2.2808
model number: 3.000000 total error :1.708497
Best Model Values: Variance: 0.8059, Bias2: 0.7136
Variance: 11.4957, Bias2: 1.1280, Total error: 12.6237
Variance: 0.5895, Bias2: 1.1955, Total error: 1.7849
Variance: 0.9781, Bias2: 1.4401, Total error: 2.4183
Variance: 0.5374, Bias2: 1.1802, Total error: 1.7175
Variance: 0.8490, Bias2: 1.4749, Total error: 2.3240
model numbe

#ANALYSIS
**Analyse the results of model performance according to different degrees of polynomial and the number of folds used. You can manipulate the code and share your analysis in terms of the performance of the model (mse and total error),
such as for instnace which degree of the model complexity (in relation to the polynomial order) would give a better model? Feel free to include other analysis about the generated models in relation to their performance results.
You can event plot the results to support your analysis.**




In [1]:

degrees = range(1, 6)  # Try polynomial degrees from 1 to 5
folds_range = range(5, 21, 5) #try fold #'s 5-20
degree_error_list = []
# Loop over different folds and degrees
for fold in folds_range:
    for degree in degrees:
        mse, best_model, total_error = polynomial_regression(degree, X, y, folds=fold, random_state=42)
        degree_error_list.append({"Fold": fold, "Degree": degree, "Total Error": total_error, "MSE": mse})

# Convert results to DataFrame for easy visualization
degree_error_df = pd.DataFrame(degree_error_list)

print(degree_error_df.head())

# Plot MSE vs. Polynomial Degree
plt.figure()
for fold in folds_range:
    fold_data = degree_error_df[degree_error_df["Fold"] == fold]
    plt.plot(fold_data["Degree"], fold_data["MSE"], marker='o', label=f'Fold {fold}')

plt.title('MSE vs. Polynomial Degree')
plt.xlabel('Polynomial Degree')
plt.ylabel('MSE')
plt.legend()
plt.grid(True)
plt.show()

# Plot Total Error vs. Polynomial Degree for each fold
plt.figure()
for fold in folds_range:
    fold_data = degree_error_df[degree_error_df["Fold"] == fold]
    plt.plot(fold_data["Degree"], fold_data["Total Error"], marker='o', label=f'Fold {fold}')

plt.title('Total Error vs. Polynomial Degree')
plt.xlabel('Polynomial Degree')
plt.ylabel('Total Error')
plt.legend()
plt.grid(True)
plt.show()


NameError: name 'polynomial_regression' is not defined

One interesting pattern I observed with the default k-fold validation settings was that the first split in each degree had significantly higher variance and bias compared to subsequent splits, especially for higher-degree models. This pattern was likely due to the presence of outliers in the first fold or the model overfitting to the training set. When I set the shuffle parameter to True, however, this issue was remedied, as the outliers were evenly distributed across all folds. However, this resulted in a higher error throughout the other folds. To prevent the outliers from affecting all folds, I set shuffle back to False, although usually this wouldn't be the case, since outliers would typically be handled through proper exploratory data analysis (EDA).

Analyzing the results with 10-fold validation, the best model in terms of mean squared error (MSE) was the polynomial model of degree 5 (MSE = 0.59), followed closely by the model of degree 4 (MSE = 0.60). However, when considering the total error (variance + bias squared), the degree 1 polynomial model had the lowest total error (1.1607). This highlights the trade-off between model complexity and overfitting. Higher-degree models perform better on MSE, they may not generalize as well, with more variance between predicted values and more variance between predicted and actual values when run on the test set. 


As seen from the plots above, MSE decreased with an increasing number of folds. More folds provide more data for training, leading to models which better represent the entire dataset. This pattern held for the total error as well, though there were diminishing returns beyond 15 folds, with 20 folds barely being an improvement over the previous iteration (red line [20 folds] vs. green line [15 folds]). Higher numbers of folds also seem to even out the differential in error between the higher-degree values; while with 5-10 folds, 5 degrees seems to be a significant improvement over 4 degrees, that distance in total error values is less pronounced when there are more folds. On the other hand, MSE clearly decreases with higher degrees, no matter the number of folds.

